src/verify_cert.rs

```diff
@@ -767,36 +803,43 @@ mod tests {
 
         let ee_der = make_end_entity(&issuer);
         let ee_cert = EndEntityCert::try_from(&ee_der).unwrap();
-        let path = match verify_chain(anchors, &intermediates, &ee_cert, None) {
-            Ok(path) => path,
-            Err(err) => return Err(err),
-        };
 
-        assert_eq!(path.anchor().subject, anchor.subject);
-        assert_eq!(path.end_entity().subject, ee_cert.subject);
-        assert_eq!(path.intermediate_certificates().count(), chain_length);
-
-        let intermediate_certs = intermediates
-            .iter()
-            .map(|der| Cert::from_der(untrusted::Input::from(der.as_ref())).unwrap())
-            .collect::<Vec<_>>();
-
-        for (cert, expected) in path
-            .intermediate_certificates()
-            .rev()
-            .zip(intermediate_certs.iter())
-        {
-            assert_eq!(cert.subject, expected.subject);
-        }
+        let expected_chain = |path: &VerifiedPath<'_>| {
+            assert_eq!(path.anchor().subject, anchor.subject);
+            assert_eq!(path.end_entity().subject, ee_cert.subject);
+            assert_eq!(path.intermediate_certificates().count(), chain_length);
+
+            let intermediate_certs = intermediates
+                .iter()
+                .map(|der| Cert::from_der(untrusted::Input::from(der.as_ref())).unwrap())
+                .collect::<Vec<_>>();
+
+            for (cert, expected) in path
+                .intermediate_certificates()
+                .rev()
+                .zip(intermediate_certs.iter())
+            {
+                assert_eq!(cert.subject, expected.subject);
+            }
 
-        for (cert, expected) in path
-            .intermediate_certificates()
-            .zip(intermediate_certs.iter().rev())
-        {
-            assert_eq!(cert.subject, expected.subject);
-        }
+            for (cert, expected) in path
+                .intermediate_certificates()
+                .zip(intermediate_certs.iter().rev())
+            {
+                assert_eq!(cert.subject, expected.subject);
+            }
 
-        Ok(())
+            Ok(())
+        };
+
+        verify_chain(
+            anchors,
+            &intermediates,
+            &ee_cert,
+            Some(&expected_chain),
```

Comment:
WDYT about also adding a test that a callback returning an err works as expected?

Reply:
Do you have a suggestion for a test case for this? I've been unable to come up with a scenario for it.

Reply:
@djc How about this approach: https://github.com/rustls/webpki/pull/191

It's a little bit complicated to arrange the certificates but seems to work. (Inspired by Sleevi's [chain of pain](https://medium.com/@sleevi_/path-building-vs-path-verifying-the-chain-of-pain-9fbab861d7d6) post).